src/catalog/src/durable/error.rs

```diff
@@ -119,3 +130,102 @@ impl From<TryFromProtoError> for DurableCatalogError {
         DurableCatalogError::Proto(e)
     }
 }
+
+/// An error that indicates the durable catalog has been fenced.
+///
+/// The order of this enum indicates the most information to the least information.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, thiserror::Error)]
+pub enum FenceError {
+    /// This instance was fenced by another instance with a higher deployment generation. This
+    /// necessary means that the other instance also had a higher epoch. The instance that fenced
+    /// us believes that they are from a later generation than us.
+    #[error("current catalog deployment generation {current_generation} fenced by new catalog epoch {fence_generation}")]
+    DeployGeneration {
+        current_generation: u64,
+        fence_generation: u64,
+    },
+    /// This instance was fenced by another instance with a higher epoch. The instance that fenced
+    /// us may or may not be from a later generation than us, we were unable to determine.
+    #[error("current catalog epoch {current_epoch} fenced by new catalog epoch {fence_epoch}")]
+    Epoch {
+        current_epoch: Epoch,
+        fence_epoch: Epoch,
+    },
+    // It's very likely that if we were to read the latest updates after this error, then we'd see
+    // a higher epoch/deploy generation from the other writer and get fenced through that. We could
+    // potentially always do that and eliminate this fence variant. However, the catalog debug tool
+    // is able to write to the catalog without incrementing the epoch and therefore would not be
+    // able to fence this instance. It's technically possible for this instance to survive and
+    // respond to a write from the catalog debug tool, but until more thought is put into it, the
+    // safest thing to do is crash and reboot.
+    /// This instance was fenced while trying to write to the catalog by some other writer.
+    #[error(
+        "current catalog upper {expected_upper:?} fenced by new catalog upper {actual_upper:?}"
+    )]
+    Upper {
+        expected_upper: Timestamp,
+        actual_upper: Timestamp,
+    },
+    /// This instance was fenced while writing to the migration shard during 0dt builtin table
+    /// migrations.
+    #[error(
+        "builtin table migration shard upper {expected_upper:?} fenced by new builtin table migration shard upper {actual_upper:?}"
+    )]
+    Migration {
```

Comment:
ignore if you got better things to do, but this could be `MigrationUpper` to match `Upper`